layers/dhcpv4.go

```diff
@@ -0,0 +1,426 @@
+	dhcp.ServerName = data[44:108]
+	dhcp.File = data[108:236]
+	if binary.BigEndian.Uint32(data[236:240]) != dhcpMagic {
+		return errors.New("Bad DHCP header")
+	}
+
+	if len(data) > 240 {
+		options := make([]byte, len(data)-240)
+		if binary.Read(bytes.NewBuffer(data[240:]), binary.BigEndian, &options) != nil {
+			return errors.New("failed to unmarshal options")
+		}
+
+		stop := len(options)
+		start := 0
+		for start < stop {
+			o := DHCPOption{}
+			if err := o.Unmarshal(options[start:]); err != nil {
+				return err
+			}
+			if o.Type == DHCP_OPT_END {
+				return nil
+			}
+			dhcp.Options = append(dhcp.Options, o)
+			start += int(o.Length) + 2
+		}
+	}
+	return nil
+
+}
+
+func (dhcp *DHCPv4) Len() uint16 {
+	n := uint16(240)
+	for _, o := range dhcp.Options {
+		n += uint16(o.Length) + 2
+	}
+	n += 1 // for opt end
+	return n
+}
+
+// SerializeTo writes the serialized form of this layer into the
+// SerializationBuffer, implementing gopacket.SerializableLayer.
+// See the docs for gopacket.SerializableLayer for more info.
+func (dhcp *DHCPv4) SerializeTo(b gopacket.SerializeBuffer, opts gopacket.SerializeOptions) error {
+	plen := int(dhcp.Len())
+	if plen < 300 {
+		plen = 300
+	}
+
+	data, err := b.PrependBytes(plen)
+	if err != nil {
+		return err
+	}
+
+	data[0] = byte(dhcp.Operation)
+	data[1] = dhcp.HardwareType
+	data[2] = dhcp.HardwareLen
+	data[3] = dhcp.HardwareOpts
+	binary.BigEndian.PutUint32(data[4:8], dhcp.Xid)
+	binary.BigEndian.PutUint16(data[8:10], dhcp.Secs)
+	binary.BigEndian.PutUint16(data[10:12], dhcp.Flags)
+	copy(data[12:16], dhcp.ClientIP.To4())
+	copy(data[16:20], dhcp.YourIP.To4())
+	copy(data[20:24], dhcp.ServerIP.To4())
+	copy(data[24:28], dhcp.GatewayIP.To4())
+	copy(data[28:44], dhcp.ClientHWAddr)
+	copy(data[44:108], dhcp.ServerName)
+	copy(data[108:236], dhcp.File)
+	binary.BigEndian.PutUint32(data[236:240], dhcpMagic)
+
+	if len(dhcp.Options) > 0 {
+		options := make([]byte, plen-240)
+		start := 0
+		for _, o := range dhcp.Options {
+			buffer, err := o.Marshal()
+			if err != nil {
+				return err
+			}
+			copy(options[start:], buffer)
+			start += len(buffer)
+		}
+		optend := NewDHCPOption(DHCP_OPT_END, nil)
+		buffer, err := (&optend).Marshal()
+		if err != nil {
+			return err
+		}
+		copy(options[start:], buffer)
+		buf := new(bytes.Buffer)
+		if err := binary.Write(buf, binary.BigEndian, options); err != nil {
+			return err
+		}
+		copy(data[240:], buf.Bytes())
+	}
+
+	return nil
+}
+
+func (dhcp *DHCPv4) CanDecode() gopacket.LayerClass {
+	return LayerTypeDHCPv4
+}
+
+func (dhcp *DHCPv4) NextLayerType() gopacket.LayerType {
+	return gopacket.LayerTypePayload
+}
+
+func decodeDHCPv4(data []byte, p gopacket.PacketBuilder) error {
+	dhcp := &DHCPv4{}
+	err := dhcp.DecodeFromBytes(data, p)
+	if err != nil {
+		return err
+	}
+	p.AddLayer(dhcp)
+	return p.NextDecoder(gopacket.LayerTypePayload)
+}
+
+func (p *DHCPv4) String() string {
```

Comment:
Is it really needed ?
Why not rely on the standard generic function ?
Returning some other strings will break formating in tool like pcapdump for example